### PR TITLE
Transformers and estimators preserve custom index

### LINKED ISSUE #1723: TargetEncoder: preserve custom row index
The TargetEncoder currently calls:
```
        if isinstance(X, pd.DataFrame):
            X.reset_index(drop=True, inplace=True)
        if isinstance(y, pd.Series):
            y.reset_index(drop=True, inplace=True)
```
Is this what we want? Previously, there were other components such as the Imputer where resetting the index caused issues.

@freddyaboulton I see that there was some discussion on https://github.com/alteryx/evalml/pull/1401#discussion_r523162379; I'd love to hear your thoughts!

### PR DESCRIPTION
**Pull request by freddyaboulton**
### Pull Request Description

Fixes #1723 

-----
*After creating the pull request: in order to pass the **release_notes_updated** check you will need to update the "Future Release" section of* `docs/source/release_notes.rst` *to include this pull request by adding :pr:`123`.*

### CODE REVIEW

File: evalml/pipelines/components/transformers/encoders/target_encoder.py [hunk 1]
```diff
@@ -60,10 +60,6 @@ def __init__(self,
                          random_seed=random_seed)
 
     def fit(self, X, y):
-        if isinstance(X, pd.DataFrame):
```
Comment (angela97lin) [inline evalml/pipelines/components/transformers/encoders/target_encoder.py:63]: 🥳

**Review by angela97lin — APPROVED** [review]
This looks great, nice and clean tests! 🥳

**Review by jeremyliweishih — APPROVED** [review]
LGTM

**Review by chukarsten — APPROVED** [review]
Love it.

**Pull request merged**